### ISSUE TITLE
Expose grid connection limits

### ISSUE THREAD

**Issue by leandro-lucarella-frequenz** [frequenz-floss/frequenz-sdk-python#360]
### What's needed?

We need to expose the new [microgrid grid connection rated fuse maximum current](https://github.com/frequenz-floss/frequenz-api-microgrid/blob/2beb6745b2e0be82f12e8ef171bec0ed1c51362a/proto/frequenz/api/microgrid/grid.proto#L25).

For now this limit can be static (queried from the microgrid API and kept immutable). A change in the grid connection limit will need a SDK restart. In the future we want to support dynamically changing this limit.

### Low-level

Add a new `Component` that wraps the `Grid` object and its metadata, as it is done with other microgrid components.

### High-level

1. Add an object to represent the `GridConnection` and a read-only rated fuse current property
2. Add a global instance exposed via `frequenz.sdk.microgrid.grid_connection`
3. Initialize the global instance in the `initialize()` function
   a. Until the new version of the microgrid API is deployed, just fill the instance `fuse.current.rate` with `None`

We should probably also add some utility methods for checking and adjusting a current value to the accepted limits. We could accept multiple current values so it is easy to use even with 3-phase samples.

```py
class Current:

    rate: float | None

    def is_within_bounds(self, *current: float) -> bool:
        ...

    @overload
    def adjust(self, /, current: float) -> float:
        ...

    @overload
    def adjust(self, /, current1: float, current2: float, *extra_currents: float) -> list[float]:
        ...

    def adjust(self, *current: float) -> float | list[float]:
        """Truncate any current above the rated_fuse_current."""

class Fuse:

    current: Current

class GridConnection:

    fuse: Fuse
```

### Use cases

Probably the `PowerManager`/`PowerArbitrator` (#161) or the should check these limits when applying charge/discharge commands, but others might as well, like the `EVChargerPool` might want to check the limit before increasing the bounds to allow vehicles to charge with high power.

Example usage:

```python
from frequenz.sdk import microgrid

desired_current = 18.5
grid_fuse = microgrid.grid_connection.fuse
if not grid_fuse.current.is_within_bounds(desired_current):
    rate = grid_fuse.current.rate
    _logger.warning("The desired current ({desired_current}) is out of bounds (the fuse is rated {rate}). Will be adjusted.")
good_current = grid_fuse.current.adjust(desired_current)
_logger.info("Using current: {good_current}")
```

### Additional context

Design discussion:
* https://github.com/frequenz-floss/frequenz-sdk-python/discussions/322

Future work:

- [ ] Initialize with the proper microgrid API calls after the new microgrid API exposing the limit is deployed
- [ ] Support dynamically changing this limit. There are 2 main approaches for this:

  * Streaming interface: This means we need a streaming interface from the microgrid API and we should make the `rated_fuse_current` a channel that streams changes in this property, like the `battery_pool()`, etc.
  * Polling: We just add a task to keep the `fuse.current.rate` updated by polling the microgrid API (or uses a streaming gRPC call if available).

  It seems like for now the polling interface should be enough.

**Comment by thomas-nicolai-frequenz** [comment]
> This means we need a streaming interface from the microgrid API

Can't we go with polling first if thats easier then implementing a streaming interface in the API?

**Comment by leandro-lucarella-frequenz** [comment]
Let's keep discussing in https://github.com/frequenz-floss/frequenz-sdk-python/discussions/322 so we keep the issue clean.

**Comment by llucax** [comment]
Blocked on #412

**Comment by llucax** [comment]
Unblocked because #412 was merged.

**Comment by tiyash-basu-frequenz** [comment]
I see the term `GridConnection` used here. We already use `Connection` as a term in the context of the Microgrid API. `GridConnection` could be misinterpreted as a special type of `Connection`. Therefore, in order to keep the terminology unambiguous, I'd propose using `GridConnectionPoint` instead.

**Comment by llucax** [comment]
I'm not sure, I don't have a strong opinion but will play devil's advocate here, where the devil is @thomas-nicolai-frequenz :D. He said several times that we shouldn't couple the SDK with the microgrid API. IMHO in the context of the SDK the most clear term is `GridConnection`, the `Point` doesn't seem to be adding much and making a long symbol name even longer.

That aside, the class will be mostly hidden to the user, the public interface will be `microgrid.grid_connection `, but I guess the discussion applies to it too.

@frequenz-floss/python-sdk-team not sure if anybody else have something to share about this.

**Comment by shsms** [comment]
I would even shorten it and say `Grid` and `microgrid.grid.fuse`, etc,  so it is easy to type and read, etc.

**Comment by llucax** [comment]
For me `microgrid.grid` looks a bit weird. Is not a *grid* inside the microgrid (as with most other elements, like the `battery_pool`, which is something inside the microgrid), but a connection to the external world, so IMHO the `connection` part is important. But I can live with just `grid` too if everybody else thinks is clear enough.

**Comment by tiyash-basu-frequenz** [comment]
To me, `Grid`/`grid` is a clearer term than `GridConnection` :)

**Comment by shsms** [comment]
I think if we say `frequenz.sdk.grid`, it would be ambiguous.  But for `frequenz.sdk.microgrid.grid`,  I think it is clear it is in the context of the microgrid.  We could still take a day after the rc is released, just to discuss the public API.  By then we'll hopefully have some more context about how things will feel for users,  and hopefully new ideas.

**Comment by thomas-nicolai-frequenz** [comment]
Please be aware that a microgrid can have:
1. One grid connection (usually the default)
2. Multiple grid connections
3. No grid connection at all (island)

Usually the grid connection is also referenced as the metering point. Not sure if that helps but I think Luca has a point when saying a grid within the microgrid sounds a bit odd. I also feel just referencing it as `grid` might be too ambiguous. So my vote would also go along the lines of microgrid.grid_connection(s) or grid_point(s).

**Comment by tiyash-basu-frequenz** [comment]
Since we are adding a singleton here, we want to support just one grid connection point here, right? 

Also, should we error out when there is no grid connection in a microgrid?

**Comment by thomas-nicolai-frequenz** [comment]
> Also, should we error out when there is no grid connection in a microgrid?

No we can't do that. You can't kill the island scenario. Or am I misunderstanding something here?

**Comment by tiyash-basu-frequenz** [comment]
> No we can't do that. You can't kill the island scenario. Or am I misunderstanding something here?

I will include the island scenario.

**Comment by llucax** [comment]
> Please be aware that a microgrid can have:
>
 >   One grid connection (usually the default)
 >   Multiple grid connections
 >   No grid connection at all (island)

Here we are talking about an abstraction, like the battery pool. It will be a "virtual" connection, for example when we need to check if we are surpassing the allowed current accepted buy the fuses of multiple grid connections, we don't want the user to know the topology and have to do the math themselves.

**Comment by llucax** [comment]
About the island scenario, shouldn't in that case just be `microgrid.grid == None`?

**Comment by tiyash-basu-frequenz** [comment]
> About the island scenario, shouldn't in that case just be `microgrid.grid == None`?

Correct.

**Comment by llucax** [comment]
As a downside the type of `microgrid.grid` will have to be *optional* (`Grid | None`) and each time one wants to access `microgrid.grid` it will have to do an `assert microgrid.grid is not None` to make the type checker happy. Although I guess one can do:

```py
grid = microgrid.grid
if grid is None:
    return  # for example, if what we are going to do needs a grid connection
# Here thanks to type narrowing grid has type `Grid` and can be used normally
```

The good thing is that users are forced to consider the island case, the bad thing is that the most common case (not an island) will need to be a bit extra verbose.

**Comment by tiyash-basu-frequenz** [comment]
Does it make sense to abstract `Fuse` further? Note that fuses will not report current metrics live. The only property it has is a rated max current, and this property is a static value.

One thing that we could do right now is add a limit per-phase to the `Fuse` class, so the upper-layers remain unaffected when we start supporting fuses in the Microgrid API.
